### PR TITLE
OrderedRDD2 take 2

### PR DESCRIPTION
**Pull request by cseed**
Builds on: https://github.com/hail-is/hail/pull/2228

It is smaller, cleaner and more self-contained, but I could still break it into more pieces if needed.

Some remarks:
 - I removed the BroadcastTypeTree nonsense.  This will kill KeyTable joins but we can fix that later.
 - I sample keys as part of collecting the partition key info.
 - I left off two features off the key ranges sampler vs OrderedRDD: I don't resample large partitions and I don't use weights to calculate the ranges.
 - I included the minimal LoadVCF interface changes.  VCF parser will come as a separate PR.

### CODE REVIEW

**Comment by danking** [comment]
@cseed looks like something is wrong I the docs / python tests.

**Comment by cseed** [comment]
It was a bug in the verify biallelic that I fixed yesterday.  I'll rebase.

File: src/main/scala/is/hail/annotations/ReadRowsRDD.scala [hunk 1]
```diff
@@ -1,21 +1,49 @@
...
 import is.hail.utils.{SerializableHadoopConfiguration, _}
-import net.jpountz.lz4.{LZ4BlockInputStream, LZ4Factory}
+import is.hail.variant.LZ4Utils
 import org.apache.commons.lang3.StringUtils
 import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.Row
 import org.apache.spark.{Partition, SparkContext, TaskContext}
 
-class Decoder(in: DataInputStream) {
+final class Decoder() {
```
Comment (danking) [inline src/main/scala/is/hail/annotations/ReadRowsRDD.scala]: somewhat of an aside: should `readRegionValue` be called `readCompoundData`? Isn't a region value just a memory and an offset? `readRegionValue` seems to be a method for reading non-primitive types.

File: src/main/scala/is/hail/expr/Relational.scala [hunk 2]
```diff
@@ -136,16 +155,94 @@ abstract class NewAST {
...
+
+            rv.offset = rvb.end()
+            rv
+          }
+        }))
+  }
+}
+
+case class MatrixValue(
+  typ: MatrixType,
```
Comment (danking) [inline src/main/scala/is/hail/expr/Relational.scala]: `MatrixLiteral`, AFAICT, was a `MatrixValue` with a `MatrixType`. `MatrixLiteral` now seems unnecessary.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 3]
```diff
@@ -224,8 +224,9 @@ object VSMSubgen {
 }
 
 class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata: VSMMetadata,
```
Comment (danking) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: We store the `VSMMetadata` twice, in VSM and in `MatrixValue`. Can we define VSM's metadata in terms of its `matrixType` field?

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 4]
```diff
@@ -641,15 +609,46 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
...
+            val v = ur.getAs[RK](1)
+            val va = ur.get(2)
+            val gs = ur.getAs[IndexedSeq[T]](3)
+
+            ec.setAll(localGlobalAnnotation, v, va)
+            aggregateOption.foreach(f => f(v, va, gs))
+            val newVA = f().zip(inserters)
+              .foldLeft(va) { case (va, (v, inserter)) =>
+                inserter(va, v)
+              }
```
Comment (danking) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: It's not per genotype, but it seems we're generally always using `while`-loops these days.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 5]
```diff
@@ -834,14 +833,44 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
...
           "vds" -> (1, other.vaSignature)))
         Annotation.buildInserter(annotationExpr, vaSignature, ec, Annotation.VARIANT_HEAD)
-      } else insertVA(other.vaSignature, Parser.parseAnnotationRoot(annotationExpr, Annotation.VARIANT_HEAD))
+      } else
+        insertVA(other.vaSignature, Parser.parseAnnotationRoot(annotationExpr, Annotation.VARIANT_HEAD))
+
+    val newMatrixType = matrixType.copy(metadata = matrixType.metadata.copy(vaSignature = finalType))
+    val newRowType = newMatrixType.rowType
+
+    // FIXME va2
```
Comment (danking) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: ???

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 6]
```diff
@@ -1027,8 +1048,19 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
       }.writeTable(path, hc.tmpDir, names.map(_.mkString("\t")), parallelWrite = parallel)
   }
 
-  def filterVariants(p: (RK, Annotation, Iterable[T]) => Boolean): VariantSampleMatrix[RPK, RK, T] =
-    copy(rdd = rdd.filter { case (v, (va, gs)) => p(v, va, gs) })
+  def filterVariants(p: (RK, Annotation, Iterable[T]) => Boolean): VariantSampleMatrix[RPK, RK, T] = {
+    val localRowType = matrixType.rowType
+    copy2(rdd2 = rdd2.filter { rv =>
+      // FIXME ur could be allocate once and set
```
Comment (danking) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: this applies to the above mapPartitions steps as well, right?

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala [hunk 7]
```diff
@@ -1391,9 +1400,46 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
...
+            rvb.addAnnotation(newRowType.fieldType(2), newVA)
+            rvb.addAnnotation(newRowType.fieldType(3), gs)
+            rvb.endStruct()
+
+            rv.offset = rvb.end()
+
+            rv
+          }
+        }))
+    /*
```
Comment (danking) [inline src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: there's a few commented out blocks like this in this file.

File: src/main/scala/is/hail/variant/VariantSampleMatrix.scala (hunk 4 above)
Comment (danking) [reply src/main/scala/is/hail/variant/VariantSampleMatrix.scala]: I believe the snippet from `ec.setAll(...)` to `val newVA = ...` can be the `f` to a `mapAnnotations` call.

File: src/main/scala/is/hail/variant/VariantDataset.scala [hunk 8]
```diff
@@ -347,51 +326,103 @@ class VariantDatasetFunctions(private val vds: VariantDataset) extends AnyVal {
...
-      else
-        Genotype(g._unboxedGT, g._fakeRef)
-    }
+    val localNSamples = vds.nSamples
+
+    val rowType = vds.matrixType.rowType
+    val gst = TArray(TGenotype)
+    val gt = TGenotype.representation
+
+    vds.copy2(
```
Comment (danking) [inline src/main/scala/is/hail/variant/VariantDataset.scala]: note to myself: need to read this carefully.

File: src/main/scala/is/hail/variant/VariantDataset.scala [hunk 9]
```diff
@@ -41,11 +41,6 @@ object VariantDataset {
 
 class VariantDatasetFunctions(private val vds: VariantDataset) extends AnyVal {
 
-  private def requireSplit(methodName: String) {
-    if (!vds.wasSplit)
-      fatal(s"method `$methodName' requires a split dataset. Use `split_multi' or `filter_multi' first.")
-  }
-
```
Comment (danking) [inline src/main/scala/is/hail/variant/VariantDataset.scala]: you don't think this error message is helpful enough to our users to justify this method's existence?

**Comment by cseed** [comment]
Closing to rebase, address comments and possibly break up a bit more.

**Pull request closed without merging**